### PR TITLE
Store computes fees

### PR DESCRIPTION
**Pull request by rosalindstengle**
Tossing this up to make sure I'm moving in more or less the right direction. Tests mostly don't exist or don't pass - gonna keep working on those.

### CODE REVIEW

File: v1/contracts/Store.sol [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    // Sets a new Oracle fee per second.
+    //TODO change permissioning
+    function setFixedOracleFeePerSecond(uint newOracleFee) external {
+        // Oracle fees at or over 100% don't make sense.
+        require(newOracleFee < FP_SCALING_FACTOR);
+        fixedOracleFeePerSecond = newOracleFee;
+        emit SetFixedOracleFeePerSecond(newOracleFee);
+    }
+
+    function computeRegularFee(uint startTime, uint endTime, uint pfc, bytes32 identifier) external view returns (uint regularFee, uint latePenalty) {
```
Comment (mrice32) [inline v1/contracts/Store.sol]: We have a new FixedPoint library: https://github.com/UMAprotocol/protocol/blob/3c77112fe7635a4519775ec4f52a73eccc4d581c/v1/contracts/FixedPoint.sol#L17. So I would take `pfc` as a FixedPoint.Unsigned and return `regularFee` and `latePenalty` as FixedPoint.Unsigned numbers.

To use this, you'll probably need to add `using FixedPoint for FixedPoint.Unsigned;` and `using FixedPoint for uint;` at the top of this contract. You'll also need to import the `FixedPoint.sol` file.

File: v1/contracts/Store.sol [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        uint timeDiff = endTime.sub(startTime);
+
+        regularFee = pfc.mul(regularFee.sub(timeDiff));
```
[the remark this answers is not in the document]
Comment (rosalindstengle) [reply v1/contracts/Store.sol]: Right!

File: v1/contracts/Store.sol [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    // Sets a new Oracle fee per second.
+    //TODO change permissioning
+    function setFixedOracleFeePerSecond(uint newOracleFee) external {
+        // Oracle fees at or over 100% don't make sense.
+        require(newOracleFee < FP_SCALING_FACTOR);
+        fixedOracleFeePerSecond = newOracleFee;
+        emit SetFixedOracleFeePerSecond(newOracleFee);
+    }
+
+    function computeRegularFee(uint startTime, uint endTime, FixedPoint.Unsigned calldata pfc,
```
Comment (mrice32) [inline v1/contracts/Store.sol]: nit: breaking this function up this way doesn't adhere to solidity style - see https://solidity.readthedocs.io/en/v0.5.8/style-guide.html#function-declaration for a guide of how to do this.

TLDR: one way that you can match their suggestions is this:
```
    function computeRegularFee(uint startTime, uint endTime, FixedPoint.Unsigned calldata pfc, bytes32 identifier)
        external
        view 
        returns (FixedPoint.Unsigned memory regularFee, FixedPoint.Unsigned memory latePenalty)
    {
```

File: v1/contracts/StoreInterface.sol [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    // Pays Oracle fees in the margin currency, erc20Address, to the store. To be used if the margin currency is an
+    // ERC20 token rather than ETH. All approved tokens are transfered.
+    function payOracleFeesErc20(address erc20Address) external; 
+
+    // Computes the regular Oracle fees that a contract should pay for a period. `pfc` is the 
+    // "profit from corruption", or the maximum amount of margin currency that a token sponsor
+    // could extract from the contract through corrupting the
+    // price feed in their favor.
+    function computeRegularFee(uint startTime, uint endTime, FixedPoint.Unsigned calldata pfc, bytes32 identifier)
```
Comment (mrice32) [inline v1/contracts/StoreInterface.sol]: Same as above.

File: v1/test/Store.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  const identifier = web3.utils.utf8ToHex("id");
+
+  //add test final fee for test identifier
+
+  beforeEach(async function() {
+    store = await Store.new();
+  });
+
+  it("Compute fees", async function() {
+    //Set fee to 10%
```
Comment (mrice32) [inline v1/test/Store.js]: nit: prefer space between `//` and start of comment here and elsewhere.

File: v1/contracts/Store.sol [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    function payOracleFeesErc20(address erc20Address) external {
+        IERC20 erc20 = IERC20(erc20Address);
+        uint authorizedAmount = erc20.allowance(msg.sender, address(this));
+        require(authorizedAmount > 0);
+        require(erc20.transferFrom(msg.sender, address(this), authorizedAmount));
+    }
+
+    // Sets a new Oracle fee per second.
+    //TODO change permissioning
```
Comment (mrice32) [inline v1/contracts/Store.sol]: IMO, we should add permissioning to this method in this PR. WDYT?

File: v1/contracts/Store.sol [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    function setFixedOracleFeePerSecond(uint newOracleFee) external {
```
Comment (mrice32) [inline v1/contracts/Store.sol]: I think this should take in a fixed point number.

File: v1/contracts/Store.sol [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+    using SafeMath for uint;
+    using FixedPoint for FixedPoint.Unsigned;
+    using FixedPoint for uint;
+
+    uint private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
+    uint private constant FP_SCALING_FACTOR = 10**18;
+
+    uint private constant WEEKLY_DELAY_FEE = 0; //<-- governance vote?
+    private mapping(address => FixedPoint.Unsigned) finalFees;
```
Comment (mrice32) [inline v1/contracts/Store.sol]: Is there any way to set the final fee for a currency?

File: v1/contracts/Store.sol [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+contract Store {
+
+    using SafeMath for uint;
+    using FixedPoint for FixedPoint.Unsigned;
+    using FixedPoint for uint;
+
+    uint private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
+    uint private constant FP_SCALING_FACTOR = 10**18;
+
+    uint private constant WEEKLY_DELAY_FEE = 0; //<-- governance vote?
```
Comment (mrice32) [inline v1/contracts/Store.sol]: I think we probably want this to be similar to `fixedOracleFeePerSecond` (set by some specific address managed by `MultiRole`). If we were to add governance, we would just give a governance contract permission to change this.

File: v1/contracts/Store.sol [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+// An implementation of StoreInterface that can accept Oracle fees in ETH or any arbitrary ERC20 token.
+contract Store {
+
+    using SafeMath for uint;
+    using FixedPoint for FixedPoint.Unsigned;
+    using FixedPoint for uint;
+
+    uint private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
+    uint private constant FP_SCALING_FACTOR = 10**18;
```
Comment (mrice32) [inline v1/contracts/Store.sol]: Ideally, if we use the `FixedPoint` library for all of our fixed point numbers, we shouldn't need this constant anymore.

Comment (rosalindstengle) [reply v1/contracts/Store.sol]: Hmm I think there is still some scaling factor off somewhere because I ended up with test cases with way too many zeros. I will dig into it

File: v1/contracts/Withdrawable.sol [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    Contract that allows the owner to withdraw any ETH and/or ERC20 tokens that the contract holds.
+*/
+
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+
+
+contract Withdrawable is Ownable {
```
Comment (rawfalafel) [inline v1/contracts/Withdrawable.sol]: This should be in `common/contracts`, no?

Although if @ptare already started working on moving `v0` to its own branch, doesn't really matter.

**Comment by rosalindstengle** [comment]
Added in a governance role that we can sort later, moved many of the values over to fp and updated the FixedPoint library to allow comparison to uint.

File: core/migrations/6_deploy_store.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+const Store = artifacts.require("Store");
+const { getKeysForNetwork, deployAndGet, addToTdr } = require("../../common/MigrationUtils.js");
+
+module.exports = async function(deployer, network, accounts) {
+  const keys = getKeysForNetwork(network, accounts);
+
+  const store = await deployAndGet(deployer, Store, { from: keys.store });
+  await addToTdr(store, network);
+
+  //TODO set oracle fees & constants
```
Comment (mrice32) [inline core/migrations/6_deploy_store.js]: nit: add space

File: core/contracts/Store.sol [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+    using SafeMath for uint;
+    using FixedPoint for FixedPoint.Unsigned;
+    using FixedPoint for uint;
+
+    enum Roles {
+        Governance
+    }
+
+    uint private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
```
Comment (mrice32) [inline core/contracts/Store.sol]: Shouldn't this be a fixed point number since it's a percentage represented as a decimal (i.e. 0.5 => 50%)?

File: core/test/Store.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+  const identifier = web3.utils.utf8ToHex("id");
+
+  //Add test final fee for test identifier
+
+  beforeEach(async function() {
+    store = await Store.new();
+  });
+
+  it("Compute fees", async function() {
```
Comment (mrice32) [inline core/test/Store.js]: I lot of the checks below are pretty isolated from one another. WDYT about separating as many as possible into distinct test cases? We've been bad about this in the past, but it would be nice to make our tests more readable :). Even one line test cases could totally make sense!

For example, this could be its own test case:
```
    // Change fee to 20%
    newFee = { value: web3.utils.toWei("0.2", "ether") };
    await store.setFixedOracleFeePerSecond(newFee, { from: owner });

    // Run time tests again
    fees = await store.computeRegularFee(100, 101, pfc, identifier);
    assert.equal(fees.regularFee.toString(), web3.utils.toWei("0.4", "ether"));

    fees = await store.computeRegularFee(100, 110, pfc, identifier);
    assert.equal(fees.regularFee.toString(), web3.utils.toWei("4", "ether"));
```

This could also be its own test case:
```
    // Disallow endTime < startTime.
    assert(await didContractThrow(store.computeRegularFee(2, 1, 10, identifier)));
```

File: core/test/Store.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+  const owner = accounts[0];
+  const derivative = accounts[1];
+  const erc20TokenOwner = accounts[2];
+
+  const identifier = web3.utils.utf8ToHex("id");
+
+  //Add test final fee for test identifier
+
+  beforeEach(async function() {
+    store = await Store.new();
```
Comment (mrice32) [inline core/test/Store.js]: Since you're migrating the store, why not use `store.deployed()`?

File: core/test/Store.js [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+  // A deployed instance of the Store contract, ready for testing.
+  let store;
+
+  const owner = accounts[0];
+  const derivative = accounts[1];
+  const erc20TokenOwner = accounts[2];
+
+  const identifier = web3.utils.utf8ToHex("id");
+
+  //Add test final fee for test identifier
```
Comment (mrice32) [inline core/test/Store.js]: Misplaced comment or a TODO?

File: core/test/Store.js [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+  beforeEach(async function() {
+    store = await Store.new();
+  });
+
+  it("Compute fees", async function() {
+    // Set fee to 10%
+    let newFee = { value: web3.utils.toWei("0.1", "ether") };
+    let txReciept = await store.setFixedOracleFeePerSecond(newFee, { from: owner });
```
Comment (mrice32) [inline core/test/Store.js]: Is this receipt being used anywhere? Usually you'd only need to use a receipt in a test to check that a particular event fired.

File: core/test/Store.js [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    assert.equal(fees.regularFee.toString(), web3.utils.toWei("4", "ether"));
+
+    // Disallow endTime < startTime.
+    assert(await didContractThrow(store.computeRegularFee(2, 1, 10, identifier)));
+
+    // Disallow setting fees higher than 100%.
+    let highFee = { value: web3.utils.toWei("1", "ether") };
+    assert(await didContractThrow(store.setFixedOracleFeePerSecond(highFee, { from: owner })));
+
+    // TODO Check that only permitted role can change the fee
```
Comment (mrice32) [inline core/test/Store.js]: Should we add role tests in this PR or do you want to do that in a follow-up?

File: core/contracts/Store.sol [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+    function computeFinalFee(address currency) 
+        external 
+        view 
+        returns (FixedPoint.Unsigned memory finalFee) 
+    {
+        finalFee = finalFees[currency];
+    }
+    
+    /*
+     * @notice Do not call this function externally.
```
Comment (mrice32) [inline core/contracts/Store.sol]: Can you add these types of comments on all the other methods. If the method is an override of the interface, you only need a docstring at the interface level, and it can be left out for the override.

File: core/contracts/Store.sol [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        // Oracle fees at or over 100% don't make sense.
+        require(newOracleFee.isLessThan(1));
+        fixedOracleFeePerSecond = newOracleFee.value;
+        emit SetFixedOracleFeePerSecond(newOracleFee);
+    }
+
+    function setFinalFee(address currency, uint finalFee) 
+        external 
+        onlyRoleHolder(uint(Roles.Governance))
```
Comment (ptare) [inline core/contracts/Store.sol]: drive-by:
I think there's some chance that this method may be called much more frequently than `setFixedOracleFeePerSecond`, because IIRC, the final fee adjusts for the relative prices of margin currencies, which could be very volatile.

Thoughts on allowing a role separate from the governance role to set fixed fees?

Comment (rosalindstengle) [reply core/contracts/Store.sol]: You're probably right about this being accessed more. Wouldn't it still be a governance vote though?

File: core/test/Store.js (hunk 18 above)
Comment (rosalindstengle) [reply core/test/Store.js]: There's a number of more types of testing I plan to add, so I'll do them all in a new PR

File: core/contracts/Store.sol (hunk 13 above)
Comment (rosalindstengle) [reply core/contracts/Store.sol]: Yeah that does make sense. I ran into some trouble making a global variable a part of a struct, but I can mess around with that more.

File: core/contracts/Store.sol (hunk 20 above)
Comment (mrice32) [reply core/contracts/Store.sol]: Good idea!

**Comment by mrice32** [comment]
It looks like everything was added twice?

File: core/contracts/Store.sol (hunk 20 above)
Comment (mrice32) [reply core/contracts/Store.sol]: It might be by governance vote at some point in the future. But in the meantime, we would be changing this with a private key. We would probably want a way to delegate this to a "hot" key. I think this same setup might make sense for the setter of the WeeklyDelayFee and the FixedOracleFee as well. We want our "super root" key to be used as little as possible.

File: core/test/Store.js [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+
+  const identifier = web3.utils.utf8ToHex("id");
+
+  // TODO Add test final fee for test identifier
+
+  beforeEach(async function() {
+    store = await Store.new();
+  });
+
+  it("Compute fees basic check", async function() {
```
Comment (mrice32) [inline core/test/Store.js]: Thanks for splitting up these tests!

File: core/test/Store.js (hunk 15 above)
Comment (mrice32) [reply core/test/Store.js]: Bump

File: core/contracts/Store.sol (hunk 19 above)
Comment (mrice32) [reply core/contracts/Store.sol]: Bump

File: core/contracts/Store.sol [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+    using SafeMath for uint;
+    using FixedPoint for FixedPoint.Unsigned;
+    using FixedPoint for uint;
+
+    enum Roles {
+        Governance
+    }
+
+    FixedPoint.Unsigned private fixedOracleFeePerSecond; // Percentage of 10^18. E.g., 1e18 is 100% Oracle fee.
```
Comment (mrice32) [inline core/contracts/Store.sol]: I think the comment needs to be reworded - "1 is a 100% Oracle fee".

File: core/contracts/FixedPoint.sol [hunk 23]
```diff
@@ -28,11 +28,31 @@ library FixedPoint {
         return a.value > b.value;
     }
 
+    /** @dev Whether `a` is less than `b`. */
```
Comment (mrice32) [inline core/contracts/FixedPoint.sol]: nit: this and the next comment both need to be updated.

File: core/contracts/Store.sol [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+pragma experimental ABIEncoderV2;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "./StoreInterface.sol";
+import "./Withdrawable.sol";
+import "./FixedPoint.sol";
+import "./MultiRole.sol";
+
+
+// An implementation of StoreInterface that can accept Oracle fees in ETH or any arbitrary ERC20 token.
```
Comment (mrice32) [inline core/contracts/Store.sol]: Please update to the same comment style as the methods (might make sense to add an `@title`).

File: core/contracts/Store.sol (hunk 19 above)
Comment (rosalindstengle) [reply core/contracts/Store.sol]: wait like comments with the @ ? I am a little confused what's missing

File: core/test/Withdrawable.js [hunk 25]
```diff
@@ -1,89 +0,0 @@
-const { didContractThrow } = require("../../common/SolidityTestUtils.js");
```
Comment (mrice32) [inline core/test/Withdrawable.js]: Why did you delete this test?

File: core/contracts/Store.sol [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+        returns (FixedPoint.Unsigned memory finalFee) 
+    {
+        finalFee = finalFees[currency];
+    }
+
+    /**
+     * @dev Sets a new oracle fee per second
+     */ 
+    function setFixedOracleFeePerSecond(FixedPoint.Unsigned memory newOracleFee) 
+        public
```
Comment (mrice32) [inline core/contracts/Store.sol]: nit: why not `external`?

File: core/contracts/Store.sol [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+        require(newOracleFee.isLessThan(1));
+        fixedOracleFeePerSecond = newOracleFee;
+        emit SetFixedOracleFeePerSecond(newOracleFee);
+    }
+    
+    /**
+     * @dev Sets a new final fee for a particular currency
+     */ 
+    function setFinalFee(address currency, FixedPoint.Unsigned memory finalFee) 
+        public
```
Comment (mrice32) [inline core/contracts/Store.sol]: nit: Why not `external`?

File: core/contracts/FixedPoint.sol [hunk 28]
```diff
@@ -28,11 +28,31 @@ library FixedPoint {
         return a.value > b.value;
     }
 
+    /** @dev Whether `a` is greater than `b`. */
+    function isGreaterThan(Unsigned memory a, uint b) internal pure returns (bool) {
```
Comment (mrice32) [inline core/contracts/FixedPoint.sol]: If you don't want to add tests for these in this PR, can we add a TODO to add them in a follow up?

File: core/contracts/Store.sol (hunk 26 above)
Comment (rosalindstengle) [reply core/contracts/Store.sol]: making this function external would mean the param would need to be calldata not memory which upsets the compiler

Comment (mrice32) [reply core/contracts/Store.sol]: Oh right - good point.
